test/Microsoft.AspNetCore.Routing.Tests/Constraints/StringRouteConstraintTest.cs

```diff
@@ -0,0 +1,117 @@
+// Licensed under the Apache License, Version 2.0. See License.txt in the project root for license information.
+
+using Microsoft.AspNetCore.Http;
+using Microsoft.AspNetCore.Routing.Constraints;
+using Moq;
+using Xunit;
+
+namespace Microsoft.AspNetCore.Routing.Constraints
+{
+    public class StringRouteConstraintTest
+    {
+        [Fact]
+        public void StringRouteConstraintSimpleTrueWithRouteDirectionIncomingRequestTest()
+        {
+            // Arrange
+            var constraint = new StringRouteConstraint("home");
+
+            // Act
+            var values = new RouteValueDictionary(new { controller = "home" });
+
+            var match = constraint.Match(
+              httpContext: Mock.Of<HttpContext>(),
+              route: new Mock<IRouter>().Object,
+              routeKey: "controller",
+              values: values,
+              routeDirection: RouteDirection.IncomingRequest);
+
+            // Assert
+            Assert.True(match);
+        }
+
+        [Fact]
+        public void StringRouteConstraintSimpleTrueWithRouteDirectionUrlGenerationTest()
+        {
+            // Arrange
+            var constraint = new StringRouteConstraint("home");
+
+            // Act
+            var values = new RouteValueDictionary(new { controller = "home" });
+
+            var match = constraint.Match(
+              httpContext: Mock.Of<HttpContext>(),
+              route: new Mock<IRouter>().Object,
+              routeKey: "controller",
+              values: values,
+              routeDirection: RouteDirection.UrlGeneration);
+
+            // Assert
+            Assert.True(match);
+        }
+
+        [Fact]
+        public void StringRouteConstraintKeyNotFoundWithRouteDirectionIncomingRequestTest()
+        {
+            // Arrange
+            var constraint = new StringRouteConstraint("admin");
+
+            // Act
+            var values = new RouteValueDictionary(new { controller = "admin" });
+
+            var match = constraint.Match(
+                httpContext: Mock.Of<HttpContext>(),
+                route: new Mock<IRouter>().Object,
+                routeKey: "action",
+                values: values,
+                routeDirection: RouteDirection.IncomingRequest);
+
+            // Assert
+            Assert.False(match);
+        }
+
+        [Fact]
+        public void StringRouteConstraintKeyNotFoundWithRouteDirectionUrlGenerationTest()
+        {
+            // Arrange
+            var constraint = new StringRouteConstraint("admin");
+
+            // Act
+            var values = new RouteValueDictionary(new { controller = "admin" });
+
+            var match = constraint.Match(
+                httpContext: Mock.Of<HttpContext>(),
+                route: new Mock<IRouter>().Object,
+                routeKey: "action",
+                values: values,
+                routeDirection: RouteDirection.UrlGeneration);
+
+            // Assert
+            Assert.False(match);
+        }
+
```

Comment:
We're also missing cases where the route-value does have a value, but the value doesn't match